### PR TITLE
gh-87092: reduce redundancy and repetition in compiler's optimization stage

### PR DESCRIPTION
**Pull request by iritkatriel**
* replace calls to ``eliminate_empty_basic_blocks`` by an assertion that there are no empty blocks
* same for a couple of inlined copies of the ``eliminate_empty_basic_blocks`` logic 
* rename ``clean_basic_block`` --> ``remove_redundant_nops`` for clarity
* rename ``normalize_basic_block`` --> ``check_cfg`` and reduce it to only check (not remove empty blocks).
* rename ``extend_block`` --> ``inline_small_exit_blocks``. Make it inline blocks with linenos too (as it used to until bde06e1b8381f140b296a397ddd1deb1c784ff8e).
* use the ``IS_UNCONDITIONAL_JUMP_OPCODE`` macro in a couple of places instead of checking precise opcodes

* Issue: gh-87092

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -9240,40 +9243,40 @@ basicblock_has_lineno(const basicblock *bb) {
     return false;
 }
 
-/* If this block ends with an unconditional jump to an exit block,
- * then remove the jump and extend this block with the target.
+/* If this block ends with an unconditional jump to a small exit block that does
+ * not have linenos, then remove the jump and extend this block with the target.
```
Comment (markshannon) [inline Python/compile.c]: OOI, has this always only copied blocks without line numbers, or is that a recent change? 
The name is  bit misleading, and there is no reason why blocks with line numbers cannot be copied.

Comment (iritkatriel) [reply Python/compile.c]: It's a recent change: bde06e1b8381f140b296a397ddd1deb1c784ff8e

Comment (iritkatriel) [reply Python/compile.c]: But I see that your subsequent changes make it unnecessary, so I think it can be reverted.

**Comment by iritkatriel** [comment]
@markshannon buildbots all pass with this. Any further comments?

File: Python/compile.c [hunk 2]
```diff
@@ -8533,9 +8538,7 @@ remove_redundant_jumps(cfg_builder *g) {
             }
         }
     }
-    if (removed) {
-        eliminate_empty_basic_blocks(g);
-    }
+    assert(no_empty_basic_blocks(g));
```
Comment (sweeneyde) [inline Python/compile.c]: It looks like the `removed` variable is unused now.

Comment (sweeneyde) [reply Python/compile.c]: If this is a postcondition, then it's a precondition as well, right? So couldn't you theoretically add the same assertion to the top of this function and replace the `if (last != NULL)` with `assert(last != NULL)`?

Comment (iritkatriel) [reply Python/compile.c]: Looks right, let me try that.

Comment (markshannon) [inline Python/compile.c]: Even if this function cannot create empty basic blocks, an assert at the end makes no sense, as the input CFG might have empty blocks. 

Remind me, does a block with just NOPs count as empty, or is that just a block with `b_iused == 0`?

File: Python/compile.c [hunk 3]
```diff
@@ -9230,50 +9230,36 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
     return -1;
 }
 
-static bool
-basicblock_has_lineno(const basicblock *bb) {
```
Comment (markshannon) [inline Python/compile.c:9234]: We still need to propagate line numbers to returns and raises, so where did this functionality go?

**Review by markshannon — COMMENTED** [review]
A couple of questions

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: > Even if this function cannot create empty basic blocks, an assert at the end makes no sense, as the input CFG might have empty blocks.

Right, following @sweeneyde 's review I added it as a precondition in this function, so I can remove it as a postcondition.

> 
> Remind me, does a block with just NOPs count as empty, or is that just a block with `b_iused == 0`?

Just  `b_iused == 0`.
(NOPs can make it to the final bytecode for line numbers).

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c:9234]: This function was never used for that. It was added in bde06e1b8381f140b296a397ddd1deb1c784ff8e.

Comment (markshannon) [reply Python/compile.c:9234]: :+1:

**Review by markshannon — APPROVED** [review]

**Pull request merged**